src/bindings.c

```diff
@@ -1249,54 +1222,150 @@ static int init_inetd_bindings(void) {
   return 0;
 }
 
+static array_header *find_server_ipbinds(pool *p, server_rec *s) {
+  register unsigned int i;
+  array_header *ipbinds = NULL;
+
+  for (i = 0; i < PR_BINDINGS_TABLE_SIZE; i++) {
+    pr_ipbind_t *ipbind;
+
+    for (ipbind = ipbind_table[i]; ipbind != NULL; ipbind = ipbind->ib_next) {
+      if (ipbind->ib_server == s) {
+        if (ipbinds == NULL) {
+          ipbinds = make_array(p, 16, sizeof(pr_ipbind_t *));
+        }
+
+        *((pr_ipbind_t **) push_array(ipbinds)) = ipbind;
+      }
+    }
+  }
+
+  return ipbinds;
+}
+
 static unsigned int process_serveralias(server_rec *s) {
   unsigned namebind_count = 0;
   config_rec *c;
+  array_header *ipbinds;
+  pool *tmp_pool;
 
   /* If there is no ipbind already for this server, we cannot associate
    * any ServerAlias-based namebinds to it.
+   *
+   * Keep in mind that there may be multiple ipbinds pointed at this server:
+   *
+   *  <VirtualHost 1.2.3.4 5.6.7.8>
+   *    ServerAlias alias
+   *  </VirtualHost>
    */
-  if (pr_ipbind_get_server(s->addr, s->ServerPort) == NULL) {
+
+  tmp_pool = make_sub_pool(s->pool);
+  pr_pool_tag(tmp_pool, "ServerAlias Processing Pool");
+
+  ipbinds = find_server_ipbinds(tmp_pool, s);
```

Comment:
This was the other key fix: ensuring that _any/all_ IP/port bindings, pointing to a vhost config, are updated for `ServerAlias` directives, not just the first such binding found.